### PR TITLE
A blinking cursor should push only one frame (#3445)

### PR DESCRIPTION
**Pull request by abarth**
Prior to this patch, we were pushing two frames each time the cursor blinked.
In turning the cursor on or off, the markNeedsPaint call was triggering another
frame to be scheduled because we cleared a bit in the scheduler at the
beginning of the frame instead of at the end of the frame.

To implement scheduling correctly, we actually need two bits: one for
ensureVisualUpdate, which just promises to get to the end of the pipeline soon,
and scheduleFrame, which promises to get to the beginning of the pipeline soon.

(Reland)

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

**Comment by abarth** [comment]
This patch is almost the same as the earlier one.  The difference is in the widgets/bindings.dart

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 1]
```diff
@@ -342,23 +336,25 @@ abstract class SchedulerBinding extends BindingBase {
   /// callbacks registered by [addPostFrameCallback].
   void handleBeginFrame(Duration rawTimeStamp) {
     Timeline.startSync('Frame');
-    assert(!_debugInFrame);
-    assert(() { _debugInFrame = true; return true; });
+    Timeline.startSync('Begin frame');
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/binding.dart]: mismerged (two startsyncs now)

Comment (abarth) [inline packages/flutter/lib/src/scheduler/binding.dart]: Oh noes!

**Comment by Hixie** [comment]
lgtm. test?

**Comment by abarth** [comment]
The trouble with writing a test for this issue is that this is the part of the system we've mocked out in the testing harness in order to let the test drive frames.

**Pull request merged**